tools/mo/openvino/tools/mo/front/kaldi/add_reshape_transpose_around_conv_pool.py

```diff
@@ -0,0 +1,215 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+import networkx as nx
+
+from openvino.tools.mo.front.common.partial_infer.utils import int64_array
+from openvino.tools.mo.front.common.replacement import FrontReplacementPattern
+from openvino.tools.mo.front.tf.graph_utils import create_op_with_const_inputs, create_op_node_with_second_input
+from openvino.tools.mo.graph.graph import Graph, Node, Port
+from openvino.tools.mo.ops.concat import Concat
+from openvino.tools.mo.ops.elementwise import Div
+from openvino.tools.mo.ops.reshape import Reshape
+from openvino.tools.mo.ops.shape import Shape
+from openvino.tools.mo.ops.transpose import Transpose
+from openvino.tools.mo.utils.shape import node_to_get_shape_value_of_indices
+
+
+def find_max_frame_time(node: Node):
+    """
+    Find maximum time_dim among all inputs of given node
+    If one of inputs have not set time_dim (=-1), then we can't find out time_dim for given node and return None
+    """
+    in_frame_time_max = 0
+
+    for inp in node.in_ports():
+        if node.in_port(inp).disconnected():
+            continue
+        in_node = node.in_port(inp).get_source().node
+        if in_node.time_dim == -1:
+            return None
+        if in_node.time_dim > in_frame_time_max:
+            in_frame_time_max = in_node.time_dim
+
+    return in_frame_time_max
+
+
+def propagate_time_dim_through_branch(node: Node, dest_port: Port):
+    """
+    Propagate time_dim for one branch starting from given node:dest_port :
+    MemoryOffset/Splice nodes increase time_dim;
+    Convolution/Pooling decrease time_dim to 0 value because it works through the whole context;
+    other nodes don't change time_dim;
+    If we find out node with several inputs, one of which have undefined time_dim, process stops.
+    """
+    child = dest_port.node
+
+    if child.time_dim >= 0:
+        return
+
+    if child.op == "MemoryOffset":
+        child.time_dim = node.time_dim + abs(child.t)
+    elif child.op == "Splice":
+        child.time_dim = node.time_dim + len(child.context) - 1
+    elif child.op in ['Convolution', 'Pooling']:
+        child.time_dim = 0
+    elif len(child.in_edges()) > 1:
+        tmp_time_dim = find_max_frame_time(child)
+        if tmp_time_dim is not None:
+            child.time_dim = tmp_time_dim
+        else:
+            return
+    else:  # len(child.in_edges()) == 1
+        child.time_dim = node.time_dim
+
+    for p in child.out_ports():
+        if child.out_port(p).disconnected():
+            continue
+        for dest in child.out_port(p).get_destinations():
+            propagate_time_dim_through_branch(child, dest)
```

Comment:
It does not work in this case. We go through all constant/inputs, then go to next level nodes. Let's imagine we have subgraph like this:
parameter -------------------- concat
const-------2nd level node --------|
Then we take concat node and will try to calculate time_dim, but can't because we have the second input (here 2nd level node that was not calculated). In topological sort all nodes are included only once, so we can't check node later. 
My method takes branch one after another and calculate as far as we can in this moment. I don't like recursion and it can be re-written in non-recursive way, but it is ideal case, non-recursive way will look much more complex.

Reply:
please try to use nx.topological_sort rountine and create wrapper in Graph class if it works in your case

Reply:
done